### PR TITLE
backup: continue to seek regions if one range has no located leader region

### LINKED ISSUE #17168: backup: requests for subranges potentially being terminated prematurely.
### What version of TiKV are you using?

master
### What operating system and CPU are you using?

N/A

### Steps to reproduce

1. build a backup request with subranges.
2. when the first `batch_size`(8 by default) regions do not intersect with the subranges. this backup task stopped immediately. 
```
For checkpoint backup. it might enter `fineGrainedBackup`.
For new feature rely on this interface. will receive `EOF` and try backup again.
```

Although the backup client can ensure data integrity and correctness, this may result in some performance regression. Therefore, it is necessary to address this issue.

### What did you expect?
all backup subranges should be handled in one backup task.
### What did happened?
the backup task on this store will exits immediately.

### PR DESCRIPTION
**Pull request by Leavrth**
### What is changed and how it works?

Issue Number: Close #17168

What's Changed:

```commit-message
backup: continue to seek regions if one range has no located leader region
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/backup/src/endpoint.rs [hunk 1]
```diff
@@ -963,10 +963,15 @@ impl<E: Engine, R: RegionInfoProvider + Clone + 'static> Endpoint<E, R> {
                     // (See https://tokio.rs/tokio/tutorial/shared-state)
                     // Use &mut and mark the type for making rust-analyzer happy.
                     let progress: &mut Progress<_> = &mut prs.lock().unwrap();
-                    let batch = progress.forward(batch_size, request.replica_read);
-                    if batch.is_empty() {
-                        return;
-                    }
+                    let batch = match progress.forward(batch_size, request.replica_read) {
```
Comment (3pointer) [inline components/backup/src/endpoint.rs:966]: how about?
```
match progress.forward(batch_size, request.replica_read) {
    Some(batch) => (batch, progress.codec.is_raw_kv, progress.cf),
    None => return,
}
```

File: components/backup/src/endpoint.rs [hunk 2]
```diff
@@ -1141,8 +1144,9 @@ impl<E: Engine, R: RegionInfoProvider + Clone + 'static> Endpoint<E, R> {
             }
         };
         let backend = Arc::<dyn ExternalStorage>::from(backend);
-        let concurrency = self.config_manager.0.read().unwrap().num_threads;
-        self.pool.borrow_mut().adjust_with(concurrency);
+        let cur_num_threads = self.config_manager.0.read().unwrap().num_threads;
+        self.pool.borrow_mut().adjust_with(cur_num_threads);
+        let concurrency = cur_num_threads * 3;
```
Comment (zhangjinpeng87) [inline components/backup/src/endpoint.rs:1149]: Please add comment to explain why multiple 3 here.

File: components/backup/src/endpoint.rs [hunk 3]
```diff
@@ -1961,6 +1960,58 @@ pub mod tests {
...
+            KeyValueCodec::new(false, ApiVersion::V1, ApiVersion::V1),
+            engine_traits::CF_DEFAULT,
+        );
+
+        let mut ranges = Vec::with_capacity(expect.len());
+        loop {
+            let n = (rand::random::<usize>() % 2) + 1;
+            let mut r = match prs.forward(n, false) {
+                None => break,
+                Some(r) => r,
```
Comment (zhangjinpeng87) [inline components/backup/src/endpoint.rs:1993]: Should explicitly cover the Some(empty Vec) case.

**Review by zhangjinpeng87 — COMMENTED** [review]
rest LGTM

File: components/backup/src/endpoint.rs (hunk 2 above)
Comment (Leavrth) [reply components/backup/src/endpoint.rs]: Just for test

File: components/backup/src/endpoint.rs (hunk 3 above)
Comment (Leavrth) [reply components/backup/src/endpoint.rs:2004]: done

**Review by zhangjinpeng87 — APPROVED** [review]
LGTM

**Pull request merged**